experimental/alp/lib/Transforms/modulo_scheduling_pass.cpp

```diff
@@ -0,0 +1,136 @@
+//===- extract_kernel_pass.cpp - Pipeline the kernel -===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+#include "alp/Transforms/Passes.h"
+#include "alp/Transforms/PassDetail.h"
+
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/Passes.h"
+#include <unordered_map>
+
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/SCF/Transforms.h"
+#include "mlir/Dialect/SCF/Utils.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/Transforms/LoopUtils.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+
+using namespace mlir;
+
+namespace{
+  struct ModuloSchedulingPass : public ModuloSchedulingPassBase<ModuloSchedulingPass>
```

Comment:
Not reviewing the pass itself for now, this would need a bunch of .mlir tests to qualify as "in reviewable state".